lib/p5.play.js

```diff
@@ -4338,4 +3938,1362 @@ p5.prototype._warn = function(message) {
+   * zero vector to indicate that no displacement is necessary.
+   * @method collide
+   * @param {p5.CollisionShape} other
+   * @return {p5.Vector}
+   */
+  p5.CollisionShape.prototype.collide = function(other) {
+    var displacee = this, displacer = other;
+
+    // Compute a displacement vector using the Separating Axis Theorem
+    // (Valid only for convex shapes)
+    //
+    // If a line (axis) exists on which the two shapes' orthogonal projections
+    // do not overlap, then the shapes do not overlap.  If the shapes'
+    // projections do overlap on all candidate axes, the axis that had the
+    // smallest overlap gives us the smallest possible displacement.
+    //
+    // @see http://www.dyn4j.org/2010/01/sat/
+    var smallestOverlap = Infinity;
+    var smallestOverlapAxis = null;
+
+    // We speed things up with an additional assumption that all collision
+    // shapes are centrosymmetric: Circles, ellipses, and rectangles
+    // are OK.  This lets us only compare the shapes' radii to the
+    // distance between their centers, even for non-circular shapes.
+    // Other convex shapes, (triangles, pentagons) will require more
+    // complex use of their projections' positions on the axis.
+    var deltaOfCenters = p5.Vector.sub(displacer.center, displacee.center);
+
+    // It turns out we only need to check a few axes, defined by the shapes
+    // being checked.  For a polygon, the normal of each face is a possible
+    // separating axis.
+    var candidateAxes = p5.CollisionShape._getCandidateAxesForShapes(displacee, displacer);
+    var axis, deltaOfCentersOnAxis, distanceOfCentersOnAxis;
+    for (var i = 0; i < candidateAxes.length; i++) {
+      axis = candidateAxes[i];
+
+      // If distance between the shape's centers as projected onto the
+      // separating axis is larger than the combined radii of the shapes
+      // projected onto the axis, the shapes do not overlap on this axis.
+      deltaOfCentersOnAxis = p5.Vector.project(deltaOfCenters, axis);
+      distanceOfCentersOnAxis = deltaOfCentersOnAxis.mag();
+      var r1 = displacee._getRadiusOnAxis(axis);
+      var r2 = displacer._getRadiusOnAxis(axis);
+      var overlap = r1 + r2 - distanceOfCentersOnAxis;
+      if (overlap <= 0) {
+        // These shapes are separated along this axis.
+        // Early-out, returning a zero-vector displacement.
+        return new p5.Vector();
+      } else if (overlap < smallestOverlap) {
+        // This is the smallest overlap we've found so far - store some
+        // information about it, which we can use to give the smallest
+        // displacement when we're done.
+        smallestOverlap = overlap;
+        // Normally use the delta of centers, which gives us direction along
+        // with an axis.  In the rare case that the centers exactly overlap,
+        // just use the original axis
+        if (deltaOfCentersOnAxis.x === 0 && deltaOfCentersOnAxis.y === 0) {
+          smallestOverlapAxis = axis;
```

Comment:
Yet another unfortunate edge case here - in some cases (two exactly-overlapped circles, for example) you can end up checking a <0,0,0> axis, which gets returned and registers as 'not overlapping' when there's actually a ton of overlap.  Fixing...